server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java

```diff
@@ -612,4 +634,113 @@ private void failShard(ReplicationFailedException e, IndexShard indexShard) {
+            for (ShardId shardId : shardsToFail) {
+                IndexShard replicaShard = this.indicesService.getShardOrNull(shardId);
+                if (replicaShard != null && replicaShard.state() == IndexShardState.STARTED) {
+                    failShard(new ReplicationFailedException("replica too far behind primary, marking as stale"), replicaShard);
+                }
+            }
+        }
+
+        @Override
+        protected String getThreadPool() {
+            return ThreadPool.Names.GENERIC;
+        }
+
+        @Override
+        public String toString() {
+            return "fail_stale_replica";
+        }
+
+    }
+
+    /**
+     * Internal class to track the list of Replication Checkpoints which are still not processed.
+     */
+    public static class PendingCheckpoints {
+        protected final Map<ShardId, List<Tuple<ReplicationCheckpoint, Long>>> checkpointsTracker = ConcurrentCollections
+            .newConcurrentMap();
+        private volatile TimeValue maxAllowedReplicationTime;
+
+        public PendingCheckpoints(TimeValue maxAllowedReplicationTime) {
+            this.maxAllowedReplicationTime = maxAllowedReplicationTime;
+        }
+
+        public void setMaxAllowedReplicationTime(TimeValue maxAllowedReplicationTime) {
+            this.maxAllowedReplicationTime = maxAllowedReplicationTime;
+        }
+
+        public void addNewReceivedCheckpoint(ShardId shardId, ReplicationCheckpoint checkpoint) {
+            checkpointsTracker.putIfAbsent(shardId, new ArrayList<>());
+            List<Tuple<ReplicationCheckpoint, Long>> checkpoints = checkpointsTracker.get(shardId);
+            if (checkpoints.size() > 0) {
+                assert checkpoints.get(checkpoints.size() - 1).v1().isAheadOf(checkpoint) == false;
+            }
+            checkpoints.add(new Tuple<>(checkpoint, System.currentTimeMillis()));
+        }
+
+        public void updateCheckpointProcessed(ShardId shardId, ReplicationCheckpoint checkpoint) {
+            if (checkpointsTracker.containsKey(shardId)) {
+                checkpointsTracker.put(
+                    shardId,
+                    checkpointsTracker.get(shardId)
+                        .stream()
+                        .filter(t -> t.v1().isAheadOf(checkpoint))
+                        .collect(Collectors.toCollection(ArrayList::new))
+                );
+            }
+        }
+
+        public ReplicationCheckpoint getLatestReplicationCheckpoint(ShardId shardId) {
+            List<Tuple<ReplicationCheckpoint, Long>> checkpoints = checkpointsTracker.getOrDefault(shardId, Collections.emptyList());
+            if (checkpoints.size() > 0) {
+                return checkpoints.get(checkpoints.size() - 1).v1();
+            }
+            return null;
+        }
+
+        public void remove(ShardId shardId) {
+            checkpointsTracker.remove(shardId);
+        }
+
+        public List<ShardId> getStaleShardsToFail() {
```

Comment:
Can you pls add a comment here on how this is computing staleness?

I am thinking that this method should also consider if the checkpoint has just started syncing to the checkpoint or not and give it a buffer.  Let's say it just completed an expensive sync that took just under the limit, we are now penalizing the shard for the new checkpoint that just began.

With that said perhaps we need to keep it simple and only consider ongoing replication time of a single sync, the setting is called `MAX_ALLOWED_REPLICATION_TIME_SETTING`.

Reply:
ack. will add a comment.

> I am thinking that this method should also consider if the checkpoint has just started syncing to the checkpoint or not and give it a buffer. Let's say it just completed an expensive sync that took just under the limit, we are now penalizing the shard for the new checkpoint that just began.

I complicates the logic a bit. But maybe we can explore logging the start time for ongoing sync as well. And not fail if that ongoing sync has been going on for  X%(let's say 30) of the threshold. 


> With that said perhaps we need to keep it simple and only consider ongoing replication time of a single sync, the setting is called MAX_ALLOWED_REPLICATION_TIME_SETTING.

Had thought about that as well but essentially we are just time bounding a single replication event and not really looking at staleness of a replica. Very different from the existing solution on primary. 
Let's discuss, If keeping it simple helps, we can explore this route as well.